### ISSUE TITLE
Minor KTO mis-estimation of KL term on odd batch sizes

### ISSUE THREAD

**Issue by Reichenbachian** [huggingface/trl#1911]
In kto_trainer.py, trl calculates the KL term by reversing the labels. That'll usually work, except on odd batch sizes, in which case the middle will be a match.

```
def _get_kl_dataset(batch: Dict[str, List[Any]]) -> Dict[str, List[Any]]:
    """Creates mismatched pairs of prompts and completions for the KL dataset by reversing the order of completions."""
    batch["answer_input_ids"] = batch["answer_input_ids"][::-1]
    batch["answer_attention_mask"] = batch["answer_attention_mask"][::-1]
    return batch
```

The halo repo does it with an offset, which works with any >1 odd batch sizes.
```
                    # for estimating the KL term, match up x and y' that are not corresponding input-output pairs in the data
                    # for x_i, get a mismatched y' by just picking the subsequent y_{i+1} in the batch (desirable/undesirable status does not matter)
                    # the respective input IDs, attention mask, and so on will be prefixed by the term KL
                    indices = list(range(1, len(batch))) + [0]
                    for i in range(len(batch)):
                        batch[i].update(self.tokenize_batch_element(
                            batch[i]['prompt_text'],
                            batch[indices[i]]['target_text'],
                            batch[i]['truncation_mode'],
                            prefix='KL'
                        ))
```

`https://github.com/ContextualAI/HALOs/blob/main/dataloader.py`

Not a huge bug, but thought it'd be worth pointing out
